<a href="https://colab.research.google.com/github/AsmaBenHmidene/food-sales-predictions/blob/main/Project_1_Part_2_(Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [38]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [39]:
import pandas as pd
import numpy as np

In [40]:
filename = '/content/drive/MyDrive/Coding Dojo/Week2/sales_predictions_2023 (1).csv'
df = pd.read_csv(filename)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [41]:
df.shape

(8523, 12)

In [42]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


#null value :
item weight and outlet size

In [43]:
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

#Duplicates:
no duplicates

In [44]:
df.duplicated().sum()

0

#Irrelevant data for sales of the product in the particular store:
Item_Identifier, item_weight, outlet_identifier, oulet_establishment_year

In [45]:
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

In [46]:
df = df.drop(columns=['Outlet_Establishment_Year', 'Item_Identifier', 'Outlet_Identifier', 'Item_Weight'])
df.head()

Item_Fat_Content  Item_Visibility              Item_Type  Item_MRP  \
0          Low Fat         0.016047                  Dairy  249.8092   
1          Regular         0.019278            Soft Drinks   48.2692   
2          Low Fat         0.016760                   Meat  141.6180   
3          Regular         0.000000  Fruits and Vegetables  182.0950   
4          Low Fat         0.000000              Household   53.8614   

  Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0      Medium               Tier 1  Supermarket Type1          3735.1380  
1      Medium               Tier 3  Supermarket Type2           443.4228  
2      Medium               Tier 1  Supermarket Type1          2097.2700  
3         NaN               Tier 3      Grocery Store           732.3800  
4        High               Tier 3  Supermarket Type1           994.7052

check if Item_Visibility is relevent data

In [47]:
filter_visibility= df['Item_Visibility'] != 0.00
df_filter = df.loc[filter_visibility, :]
df_filter.head()

Item_Fat_Content  Item_Visibility    Item_Type  Item_MRP Outlet_Size  \
0          Low Fat         0.016047        Dairy  249.8092      Medium   
1          Regular         0.019278  Soft Drinks   48.2692      Medium   
2          Low Fat         0.016760         Meat  141.6180      Medium   
6          Regular         0.012741  Snack Foods   57.6588        High   
7          Low Fat         0.127470  Snack Foods  107.7622      Medium   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
6               Tier 3  Supermarket Type1           343.5528  
7               Tier 3  Supermarket Type3          4022.7636

In [48]:
df_filter['Item_Visibility'].describe().round(2)

count    7997.00
mean        0.07
std         0.05
min         0.00
25%         0.03
50%         0.06
75%         0.10
max         0.33
Name: Item_Visibility, dtype: float64

In [49]:
df_filter['Item_Visibility'].idxmax()

854

In [50]:
df_filter.iloc[854]['Item_Type']

'Fruits and Vegetables'

In [51]:
df_filter['Item_Outlet_Sales'].idxmax()

7188

In [52]:
df_filter.iloc[7188]['Item_Type']

'Health and Hygiene'

In [53]:
df_filter['Item_Visibility'].idxmin()

7551

In [54]:
df_filter.iloc[7551]['Item_Type']

'Fruits and Vegetables'

In [55]:
df_filter.groupby('Item_Type')['Item_Visibility'].mean().round(2)

Item_Type
Baking Goods             0.07
Breads                   0.07
Breakfast                0.09
Canned                   0.07
Dairy                    0.08
Frozen Foods             0.07
Fruits and Vegetables    0.07
Hard Drinks              0.07
Health and Hygiene       0.06
Household                0.06
Meat                     0.07
Others                   0.06
Seafood                  0.08
Snack Foods              0.07
Soft Drinks              0.07
Starchy Foods            0.08
Name: Item_Visibility, dtype: float64

In [56]:
df_filter.groupby('Item_Type')['Item_Visibility'].mean().round(2).value_counts()

0.07    9
0.08    3
0.06    3
0.09    1
Name: Item_Visibility, dtype: int64

In [57]:
df_filter.groupby('Outlet_Type')['Item_Visibility'].mean().round(2)

Outlet_Type
Grocery Store        0.11
Supermarket Type1    0.06
Supermarket Type2    0.07
Supermarket Type3    0.06
Name: Item_Visibility, dtype: float64

Conclusion: Item_visibility is very variable according to different parameters: the most visible item is the least visible as well and is not the most selled one. we will drop it as well

In [58]:
df = df.drop(columns=['Item_Visibility'])
df.head()

Item_Fat_Content              Item_Type  Item_MRP Outlet_Size  \
0          Low Fat                  Dairy  249.8092      Medium   
1          Regular            Soft Drinks   48.2692      Medium   
2          Low Fat                   Meat  141.6180      Medium   
3          Regular  Fruits and Vegetables  182.0950         NaN   
4          Low Fat              Household   53.8614        High   

  Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
0               Tier 1  Supermarket Type1          3735.1380  
1               Tier 3  Supermarket Type2           443.4228  
2               Tier 1  Supermarket Type1          2097.2700  
3               Tier 3      Grocery Store           732.3800  
4               Tier 3  Supermarket Type1           994.7052

check if Item_MRP is relevent data

In [59]:
df['Item_MRP'].describe().round(2)

count    8523.00
mean      140.99
std        62.28
min        31.29
25%        93.83
50%       143.01
75%       185.64
max       266.89
Name: Item_MRP, dtype: float64

In [60]:
df.iloc[df['Item_MRP'].idxmax()]['Item_Type']

'Canned'

In [61]:
df.iloc[df['Item_MRP'].idxmin()]['Item_Type']

'Soft Drinks'

In [62]:
df.groupby('Item_Type')['Item_MRP'].mean().round(2)

Item_Type
Baking Goods             126.38
Breads                   140.95
Breakfast                141.79
Canned                   139.76
Dairy                    148.50
Frozen Foods             138.50
Fruits and Vegetables    144.58
Hard Drinks              137.08
Health and Hygiene       130.82
Household                149.42
Meat                     139.88
Others                   132.85
Seafood                  141.84
Snack Foods              146.19
Soft Drinks              131.49
Starchy Foods            147.84
Name: Item_MRP, dtype: float64

Conclusion: Item_MRP is not very relevent data to predict sales. we will drop it as well.

In [63]:
df = df.drop(columns=['Item_MRP'])
df.head()

Item_Fat_Content              Item_Type Outlet_Size Outlet_Location_Type  \
0          Low Fat                  Dairy      Medium               Tier 1   
1          Regular            Soft Drinks      Medium               Tier 3   
2          Low Fat                   Meat      Medium               Tier 1   
3          Regular  Fruits and Vegetables         NaN               Tier 3   
4          Low Fat              Household        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

#missing value outlet size:

In [64]:
df['Outlet_Size'].fillna('Missing', inplace=True)

In [65]:
df['Outlet_Size'].value_counts(normalize = True).mul(100).round(2)

Medium     32.77
Missing    28.28
Small      28.02
High       10.94
Name: Outlet_Size, dtype: float64

missing value represent 28.28% which is more then 2% thus we can't drop rows and less then 50% thus it is better not to drop the column.
Adding the missing value to medium category will affect the study correlation with item outlet sales. Therefore, we will keep the missing value in a category called "Missing"

In [66]:
df.groupby('Outlet_Size')['Item_Outlet_Sales'].mean().round(2)

Outlet_Size
High       2299.00
Medium     2681.60
Missing    1822.63
Small      1912.15
Name: Item_Outlet_Sales, dtype: float64

Conclusion: most sales are done in medium size outlet.

#inconsistent categories of data

In [67]:
df['Item_Fat_Content'].value_counts(normalize = True).mul(100).round(2)

Low Fat    59.71
Regular    33.90
LF          3.71
reg         1.37
low fat     1.31
Name: Item_Fat_Content, dtype: float64

In [68]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF','Low Fat').replace('low fat','Low Fat').replace('reg','Regular')
df['Item_Fat_Content'].value_counts(normalize = True).mul(100).round(2)

Low Fat    64.73
Regular    35.27
Name: Item_Fat_Content, dtype: float64

In [69]:
df.groupby('Item_Fat_Content')['Item_Outlet_Sales'].mean().round(2)

Item_Fat_Content
Low Fat    2157.71
Regular    2224.56
Name: Item_Outlet_Sales, dtype: float64

In [70]:
filter_FatContent= df['Item_Fat_Content'] == "Regular"
df_filter_FatContent = df.loc[filter_FatContent, :]
df_filter_FatContent.head()

Item_Fat_Content              Item_Type Outlet_Size Outlet_Location_Type  \
1          Regular            Soft Drinks      Medium               Tier 3   
3          Regular  Fruits and Vegetables     Missing               Tier 3   
5          Regular           Baking Goods      Medium               Tier 3   
6          Regular            Snack Foods        High               Tier 3   
8          Regular           Frozen Foods     Missing               Tier 2   

         Outlet_Type  Item_Outlet_Sales  
1  Supermarket Type2           443.4228  
3      Grocery Store           732.3800  
5  Supermarket Type2           556.6088  
6  Supermarket Type1           343.5528  
8  Supermarket Type1          1076.5986

In [71]:
df_filter_FatContent.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmax()

'Seafood'

In [72]:
df_filter_FatContent.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmin()

'Soft Drinks'

In [73]:
filter_FatContent2= df['Item_Fat_Content'] == "Low Fat"
df_filter_FatContent2 = df.loc[filter_FatContent2, :]
df_filter_FatContent2.head()

Item_Fat_Content              Item_Type Outlet_Size Outlet_Location_Type  \
0           Low Fat                  Dairy      Medium               Tier 1   
2           Low Fat                   Meat      Medium               Tier 1   
4           Low Fat              Household        High               Tier 3   
7           Low Fat            Snack Foods      Medium               Tier 3   
10          Low Fat  Fruits and Vegetables      Medium               Tier 1   

          Outlet_Type  Item_Outlet_Sales  
0   Supermarket Type1          3735.1380  
2   Supermarket Type1          2097.2700  
4   Supermarket Type1           994.7052  
7   Supermarket Type3          4022.7636  
10  Supermarket Type1          1516.0266

In [74]:
df_filter_FatContent2.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmax()

'Starchy Foods'

In [75]:
df_filter_FatContent2.groupby('Item_Type')['Item_Outlet_Sales'].mean().round(2).idxmin()

'Baking Goods'

conclusion: regular products represent the principle sales and more specifically seafood.
For Low fat item, starchy foods represent the most selled product.

#Effect of outlet type on sales

In [76]:
df['Outlet_Type'].value_counts(normalize = True).mul(100).round(2)

Supermarket Type1    65.43
Grocery Store        12.71
Supermarket Type3    10.97
Supermarket Type2    10.89
Name: Outlet_Type, dtype: float64

In [77]:
df.groupby('Outlet_Type')['Item_Outlet_Sales'].mean().round(2)

Outlet_Type
Grocery Store         339.83
Supermarket Type1    2316.18
Supermarket Type2    1995.50
Supermarket Type3    3694.04
Name: Item_Outlet_Sales, dtype: float64

Conclusion: Supermarket Type3 represent only 10.97% however they show the highest sales average.

#Effect of the oulet location type on sales

In [78]:
df['Outlet_Location_Type'].value_counts(normalize = True).mul(100).round(2)

Tier 3    39.31
Tier 2    32.68
Tier 1    28.02
Name: Outlet_Location_Type, dtype: float64

In [79]:
df.groupby('Outlet_Location_Type')['Item_Outlet_Sales'].mean().round(2)

Outlet_Location_Type
Tier 1    1876.91
Tier 2    2323.99
Tier 3    2279.63
Name: Item_Outlet_Sales, dtype: float64

Conclusion: Outlet in location Tier 2 shows the highest sales average.

#Most and least selled items

In [83]:
df.groupby('Item_Type')['Item_Outlet_Sales'].mean().idxmax()

'Starchy Foods'

In [84]:
df.groupby('Item_Type')['Item_Outlet_Sales'].mean().idxmin()

'Others'

#Final conclusion:

The properties of products and outlets that play crucial roles in increasing sales:
- Outlet_location_type: Tier 2
- Outlet_Type: Supermarket Type3
- Outlet_Size: medium size outlet
- Item_Fat_Content: regular products represent the principle sales and more specifically seafood. For Low fat item, starchy foods represent the most selled product.